### ISSUE TITLE
Initial DTLS timeout too low for high-latency connections

### ISSUE THREAD

**Issue by fancycode** [meetecho/janus-gateway#1377]
The initial DTLS timeout is 100 msec:
https://github.com/meetecho/janus-gateway/blob/7b3a9f003222c02ab410eec12ddf95656598c74d/dtls.c#L545

For high-latency connections, this is too low and will cause the connection to fail in `SSL_read` after a couple of seconds here:
https://github.com/meetecho/janus-gateway/blob/7b3a9f003222c02ab410eec12ddf95656598c74d/dtls.c#L650
```
[ERR] [dtls.c:janus_dtls_srtp_incoming_msg:658] [10521928651745] Handshake error: error:100000c6:SSL routines:OPENSSL_internal:READ_TIMEOUT_EXPIRED
```

I noticed this with connections from Australia to one of our servers in Europe with a latency of over 300 msec.

Changing the initial timeout to 1000 msec (which is the strongly recommended value from RFC 6347, see https://tools.ietf.org/html/rfc6347#section-4.2.4.1) fixes the connection issues.

Chrome uses an initial value that is based on the RTT determined during ICE negotiation:
https://cs.chromium.org/chromium/src/third_party/webrtc/p2p/base/dtlstransport.cc?l=787&rcl=d574123c5035901107e4dd56a3cd0968f174a881

Something similar should be used in Janus, too (if this is possible with libnice) - or a larger (or configurable) default value.

**Comment by lminiero** [comment]
The 100ms you see is only enforced if you use BoringSSL: with OpenSSL the value is not configurable, and is indeed 1s instead. About your suggestion, libnice has nothing to do with it: it's a DTLS handshake matter, and so is up to OpenSSL: as I said, there's no way to change it with OpenSSL (they only added something like that in 1.1.1 that is not available in the vast majority of distros), so there isn't anything we can do, statically or programmatically.

**Comment by fancycode** [comment]
Yeah, I'm using BoringSSL, sorry for not mentioning this earlier. About libnice: I understand that it's a DTLS issue and not related to libnice, but the initial value for the DTLS timeout could be based on the ICE RTT similar to what Chrome is doing instead of using a hardcoded value.

For example if the RTT of a client is 300 ms like in my example above, the initial timeout in Chrome will be 600 ms, which is sufficient to prevent timeouts. If a client has a fast connection with a RTT of 30 ms, the initial timeout will be 60 ms (and packets will be retried earlier on packet loss).

OpenSSL will not have the issue as they are always using the 1000 ms timeout.

**Comment by fancycode** [comment]
Anyway, if Janus is build with BoringSSL and users have too high latencies, connections will not work - which is the actual issue.

**Comment by lminiero** [comment]
The only RTT feedback we have comes from RTCP, and that happens after the session has been established. We don't have any way to get RTT info just from ICE, I believe: I don't think libnice exposes that but I may be wrong. As such, the best we can do is probably allow the value to be configured in janus.cfg, rather than hardcode 100ms; suboptimal for your needs, but without feedback nothing else we can do.

**Comment by lminiero** [comment]
No need, I can do that myself after lunch, it should be an easy enough fix :+1:

**Comment by oej** [comment]
As a side note, actually getting RTT from the ICE session is an interesting thought for optimising other things.

A secondary side note: We tried Janus with satellite connections (over 600 ms latency) and nothing worked. We did not go through extensive debugging - but my feeling was that the web browsers wasn't built for any such latency - including web sockets and other systems.

**Comment by lminiero** [comment]
It would be nice indeed, but libnice doesn't expose that info. It's something that should be added to that library, before we can use it, as we don't have access to the internals of the connectivity checks it makes.

As to the satellite issue, please feel free to open a new issue as soon as you have some info we can look at to try and debug the problem.

**Comment by lminiero** [comment]
Made the change: @fancycode please let me know if it fixes your issue. Notice that the default if you don't specify anything in `janus.cfg`, now, isn't 100ms but 1000ms.

**Comment by fancycode** [comment]
I didn't have time yet to run this on our servers, but the code looks good.

Perhaps change this condition to `#ifndef HAVE_DTLS_SETTIMEOUT` with an updated log message to have the same condition as below where you determine if `DTLSv1_set_initial_timeout_duration` is available:
https://github.com/meetecho/janus-gateway/blob/7a0257b2503af89dcc4097e36fdc4b45199d87d0/dtls.c#L419
